### PR TITLE
fix(file source): avoid warning for empty files too small to fingerprint

### LINKED ISSUE #1065: Only warn about small files when they are not empty
This is a follow up to https://github.com/timberio/vector/pull/863#issuecomment-544314174.

To recap, when the `file` source is configured to use the `checksum` fingerprinting strategy a _warning_ log message is emitted letting the user know the file is too small to generate a fingerprint:

```rust
warn!(message = "Ignoring file smaller than fingerprint_bytes", file = ?path);
```

This is a very welcome feature for files that are actually too small. It helps to prevent confusion around why small files aren't being tailed. Unfortunately, it can produce a lot of noise in production environments where many empty files are present.

I'm unsure if we should ignore empty files, given that it could lead to the same confusion that we experienced previously with small files. Then again, I can see the reasoning behind ignoring empty files. I'm opening this up for discussion so we can come to a conclusion about what to do in this scenario.

### PR DESCRIPTION
**Pull request by an-cor**
## Summary

This PR fixes #1065 by avoiding the `Currently ignoring file too small to fingerprint` warning for empty files.

The existing behavior emits the same warning for empty files and for non-empty files that are too small to fingerprint. Empty files can be common in production logging environments, especially with short-lived jobs or pods, so warning on them can create noisy logs without giving users actionable information.

This change keeps the existing warning behavior for non-empty files that are too small to fingerprint, but skips the warning for empty files.

## Vector configuration

No full Vector configuration was used for this change. The behavior is covered at the file-source-common fingerprinter level with a targeted unit test.

## How did you test this PR?

Tested locally on macOS with targeted file-source-common tests.

Commands run:

```bash
cargo test -p file-source-common fingerprint_or_emit
cargo test -p file-source-common fingerprinter
cargo fmt --check
git diff --check
cargo clippy -p file-source-common --all-targets
```

Results:

fingerprint_or_emit test filter passed: 3 passed, 0 failed
fingerprinter test filter passed: 10 passed, 0 failed
cargo fmt --check passed
git diff --check passed
cargo clippy -p file-source-common --all-targets passed

## Change Type

* [x] Bug fix
* [ ] New feature
* [ ] Dependencies
* [ ] Non-functional (chore, refactoring, docs)
* [ ] Performance

## Is this a breaking change?

* [ ] Yes
* [x] No

## Does this PR include user facing changes?

* [x] Yes. Please add a changelog fragment based on our [[guidelines](https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md)](https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md).
* [ ] No. A maintainer will apply the `no-changelog` label to this PR.

## References

* Closes: #1065

## Notes

This PR includes a changelog fragment:

```text
changelog.d/1065-empty-file-fingerprint-warning.fix.md
```

```md id="1l6rfl"```

## AI assistance disclosure

I used Claude Code to help draft and revise parts of this change. I reviewed the implementation, ran the tests locally, and take responsibility for the contribution.

```md id="nk653h" ```
Follow-up updates after automated review:

- Empty files are still tracked in `known_small_files` for retry/cleanup behavior.
- Warning emission is tracked separately from small-file tracking.
- If a file is first empty and later becomes non-empty but still too small to fingerprint, it emits the warning once.
- Empty gzip/decompressed streams are treated as empty and do not emit the warning.

### CODE REVIEW

**Comment by ghost** [comment]
<a target='_blank' href='https://cose.datadoghq.com/ci/pipeline-executions?query=ci_level%3Ajob+%40ci.status%3Aerror+%40git.commit.sha%3Ae8f82d1f3ac3a3282ad42b0a35eb52d0465a3076&referrer=pr-comment&start=1784525025000'><picture><source media="(prefers-color-scheme: dark)" srcset="https://static.datadoghq.com/static/images/pr_comments/software_delivery/badges_header/pipelines-warning.svg"/>Pipelines</picture></a>

## ⚠️ Warnings

🚦 **1 Pipeline job failed**

<blockquote>
<div>

<code> CLA Assistant | CLAAssistant </code> &nbsp;&nbsp;<a href="https://cose.datadoghq.com/ci/pipeline-executions?query=ci_level%3Ajob&#43;ci.job_correlation_id%3A7713f7ca6838a010&amp;referrer=pr-comment&amp;start=1784525025000" title="View in Datadog">View in Datadog</a> &nbsp; <a href="https://github.com/vectordotdev/vector/actions/runs/29720378037/job/88282032085" title="View in GitHub Actions"><picture><source media="(prefers-color-scheme: dark)" srcset="https://static.datadoghq.com/static/images/pr_comments/software_delivery/logos/github-logo-dark.svg"/>GitHub Actions</picture></a>

</div>
</blockquote>

Useful? React with 👍 / 👎

<sub>
This comment will be updated automatically if new data arrives. <br />
🔗 Commit SHA: e8f82d1 | <a href="https://docs.datadoghq.com/integrations/guide/source-code-integration/#pr-comments">Docs</a> | Give us <a href="https://forms.gle/ys6MDmdQEeC5pkzWA">feedback</a>!
</sub>

**Comment by an-cor** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by an-cor** [comment]
Thanks for the review. I updated the implementation so empty files are still inserted into `known_small_files` for retry/cleanup bookkeeping, while only suppressing `emit_file_checksum_failed` for empty files. I also updated the regression test to verify that empty files remain tracked and non-empty small files still emit the checksum warning.

**Comment by an-cor** [comment]
I have read the CLA Document and I hereby sign the CLA

File: lib/file-source-common/src/fingerprinter.rs [hunk 1]
```diff
@@ -222,7 +225,9 @@ impl Fingerprinter {
                 match error.kind() {
                     ErrorKind::UnexpectedEof => {
                         if !known_small_files.contains_key(path) {
-                            emitter.emit_file_checksum_failed(path);
+                            if file_size != Some(0) {
+                                emitter.emit_file_checksum_failed(path);
```
[the remark this answers is not in the document]
Comment (an-cor) [reply lib/file-source-common/src/fingerprinter.rs]: Thanks, this was a real edge case. I updated the implementation so warning emission is tracked separately from small-file retry/cleanup tracking. Empty files are still tracked, and if a file later becomes non-empty but remains too small to fingerprint, the checksum warning is emitted once.

File: lib/file-source-common/src/fingerprinter.rs [hunk 2]
```diff
@@ -202,9 +202,12 @@ impl Fingerprinter {
         known_small_files: &mut HashMap<PathBuf, time::Instant>,
         emitter: &impl FileSourceInternalEvents,
     ) -> Option<FileFingerprint> {
+        let mut file_size = None;
+
         let metadata = match fs::metadata(path).await {
             Ok(metadata) => {
                 if !metadata.is_dir() {
+                    file_size = Some(metadata.len());
```
[the remark this answers is not in the document]
Comment (an-cor) [reply lib/file-source-common/src/fingerprinter.rs:247]: Thanks, I updated the logic to distinguish empty streams based on fingerprint read behavior instead of relying only on on-disk metadata length. I also added coverage for an empty gzip/decompressed stream so it does not emit the checksum warning while still being tracked.

File: lib/file-source-common/src/fingerprinter.rs [hunk 3]
```diff
@@ -221,9 +261,30 @@ impl Fingerprinter {
...
+                        // empty-decompressed-gzip-stream distinguishable from
+                        // a gzip file with nonzero on-disk size. If decoding
+                        // never got that far (e.g. the raw file is too short
+                        // to even hold a compression magic header), fall back
+                        // to the on-disk size, which is accurate for
+                        // uncompressed files.
+                        let is_empty = error
+                            .get_ref()
+                            .and_then(|inner| inner.downcast_ref::<TooSmallToFingerprint>())
+                            .map_or(file_size == Some(0), |inner| inner.bytes_read == 0);
```
[the remark this answers is not in the document]
Comment (an-cor) [reply lib/file-source-common/src/fingerprinter.rs:279]: 👍
